### ISSUE TITLE
Trying to build this on Mac Catalina with Miniconda Python 3.7

### ISSUE THREAD

**Issue by fgolemo** [carla-simulator/map#25]
Hey guys, here's my setup:

- Mac OS Catalina, 
- Homebrew, 
- Miniconda 3.7

Here are my steps to reproduce

```
# install boost version 1.71.0 because that must match the boost-python version of Conda and that's currently 1.71.0
brew install https://raw.githubusercontent.com/Homebrew/homebrew-core/8d748e26ccc9afc8ea0d0201ae234fda35de721e/Formula/boost.rb

conda install boost

# now boost-python is located in ~/miniconda3/lib/cmake/boost_python-1.71.0

git clone this repo, mkdir, etc. 
cd build/ad_physics
cmake ../../ad_physics -DCMAKE_INSTALL_PREFIX=../../install/ad_physics -DBUILD_PYTHON_BINDING=ON
```

Gives me the following error that I have no idea how to fix:

<img width="627" alt="Screen Shot 2020-09-07 at 5 15 09 PM" src="https://user-images.githubusercontent.com/4822928/92417116-96dbd380-f12e-11ea-8aa8-6c1d62d8b09e.png">

In particular, I'm confused by these lines:

`* libboost_python37.dylib (3.7, Boost_PYTHON_VERSION=2.7)`
`* libboost_python37.a (3.7, Boost_PYTHON_VERSION=2.7)`

Because it sounds like it's implying that I only have boost installed for Python version 2.7 (i.e. system python as opposed to miniconda) but I clearly installed it through `conda` so it definitely matches the python version.

**Comment by fabianoboril** [comment]
Hi,

as far as I know, the current master-branch tries to build a Python binding for 2.7 and one for 3.x. It first starts with Python 2.7, and if this fails, everything fails.
https://github.com/carla-simulator/map/blob/e4bbd4843c8d8e9977b7c24acaffa6992e2ae4f7/cmake/python-binding.cmake#L35

You can try the improve_python_user_interface branch from PR #23 Which should solve your problem.

**Comment by fgolemo** [comment]
Thanks, that actually makes a lot of sense.

I've tried switching to the new branch (`git checkout improve_python_user_interface`) but I'm getting the same error from the screenshot above. After switching to the new branch, I've made sure to remove the build and install directories. 
I.e. it breaks and exists after not finding a Python 2.7 lib.

I also tried changing the search loop from
`foreach(PYTHON_VERION_SEARCH 2.7 3.5 3.6 3.7 3.8)`
to 
`foreach(PYTHON_VERION_SEARCH 3.7)`
but then I get a different error:

<img width="649" alt="Screen Shot 2020-09-09 at 12 49 55 PM" src="https://user-images.githubusercontent.com/4822928/92628554-ff47c380-f29a-11ea-9071-4b94d917505e.png">

**Comment by fgolemo** [comment]
Oh, progress:

The above error is because cmake `find_package(PythonLibs 3)` obviously doesn't find miniconda/anaconda. In order to get around that, I've modified the cmake command to include the python lib path:

`cmake ../../ad_physics -DCMAKE_INSTALL_PREFIX=../../install/ad_physics -DBUILD_PYTHON_BINDING=ON -DPYTHON_INCLUDE_DIR=$(python -c "from distutils.sysconfig import get_python_inc; print(get_python_inc())")  -DPYTHON_LIBRARY=$(python -c "import distutils.sysconfig as sysconfig; print(sysconfig.get_config_var('LIBDIR'))")`

...and in the `cmake/python-binding.cmake` file I commented out the lines
```
#    unset(PYTHON_INCLUDE_DIR)
#    unset(PYTHON_INCLUDE_DIR CACHE)
#    unset(PYTHON_LIBRARY)
#    unset(PYTHON_LIBRARY CACHE)
```

and I changed 
`find_package(PythonLibs EXACT ${PYTHON_VERION_SEARCH})`
to
`find_package(PythonLibs EXACT REQUIRED ${PYTHON_VERION_SEARCH})`
because... it's required for this to work.

Now I get further in the cmake config process but I still get an error:

<img width="655" alt="Screen Shot 2020-09-09 at 12 58 03 PM" src="https://user-images.githubusercontent.com/4822928/92629312-2226a780-f29c-11ea-9e0f-9cb18bdada29.png">

**Comment by fabianoboril** [comment]
Hi,
it seems no python-cpp files are generated (Generating result: 0). It seems that pygccxml has some issues here. Let me bring in @berndgassmann, who might know more about this.

**Comment by berndgassmann** [comment]
Hi,
from the generator output it doesn't look too bad. So the question is actually what is going wrong here?

When I run it the first time, I get the following output:
```
-- Generating python binding source code at /localdisk/users/bernd/map/build/ad_physics/python
-- Generating result: 0
-- Generator output: /localdisk/users/bernd/.local/lib/python2.7/site-packages/pygccxml/declarations/declaration.py:340: DeprecationWarning: The i_depend_on_them() method is deprecated.
Please use the declarations.get_dependencies_from_decl()function instead.

  DeprecationWarning)

INFO: file "named_tuple.py" - updated( 0.000243 seconds )

INFO: file "AdPhysicsPythonWrapper.cpp.in" - updated( 0.000650 seconds )

-- Configuring done
-- Generating done
-- Build files have been written to: /localdisk/users/bernd/map/build/ad_physics
```

When I run it a second time, I get the following instead:

```
-- Generating python binding source code at /localdisk/users/bernd/map/build/ad_physics/python
-- Generating result: 0
-- Generator output: /localdisk/users/bernd/.local/lib/python2.7/site-packages/pygccxml/declarations/declaration.py:340: DeprecationWarning: The i_depend_on_them() method is deprecated.
Please use the declarations.get_dependencies_from_decl()function instead.

  DeprecationWarning)

-- Configuring done
-- Generating done
-- Build files have been written to: /localdisk/users/bernd/map/build/ad_physics
```

So it looks as if the there are some differences in the output (the INFO messages are not printed).
Indeed it looks as if the file `/localdisk/users/bernd/map/build/ad_physics/python/AdPhysicsPythonWrapper.cpp.in` is not rewritten if already existing. Therefore, I added some code to remove the file before upfront...

So did you try to remove the build/ad_physics folder and rerun the cmake command? Maybe it's just that simple that some things were not updated after your manual adaptions...

In case the file `/localdisk/users/bernd/map/build/ad_physics/python/AdPhysicsPythonWrapper.cpp.in` is not there at all, the post processing function `post_process_python_wrapper()` throws an error; so either the file exists or the post processing is not executed properly...

Maybe the file `build/ad_physics/python/generate_python_lib.py` wasn't written properly to reflect the correct required output file ... 

Argghh, got the issue after reducing the python list in my cmake to one single and adding debug output...
then we get the generate_python_lib.py code:
```
for binding in "python35":
        post_process_python_wrapper(header_directories=header_dirs, additional_replacements={("libad_physics_python", "libad_physics_{}".format(binding))}, cpp_filename_in="AdPhysicsPythonWrapper.cpp.in", cpp_filename_out="AdPhysicsPythonWrapper-{}.cpp".format(binding))
```

which interprets each character as list entry ...

```
generate_python_wrapper(): AdPhysicsPythonWrapper.cpp.in written.
post_process_python_wrapper(): AdPhysicsPythonWrapper-p.cpp written.
post_process_python_wrapper(): AdPhysicsPythonWrapper-y.cpp written.
post_process_python_wrapper(): AdPhysicsPythonWrapper-t.cpp written.
post_process_python_wrapper(): AdPhysicsPythonWrapper-h.cpp written.
post_process_python_wrapper(): AdPhysicsPythonWrapper-o.cpp written.
post_process_python_wrapper(): AdPhysicsPythonWrapper-n.cpp written.
post_process_python_wrapper(): AdPhysicsPythonWrapper-3.cpp written.
post_process_python_wrapper(): AdPhysicsPythonWrapper-5.cpp written.
```


I'll push a fix on this.

**Comment by berndgassmann** [comment]
Hi,
I've found out when looking into the details of boost 1.72.0 (used in CARLA) is that it seems that even changing the package search for Boost::python<xy> from REQUIRED to QUIET breaks if the find_package for the Boost python component is performed with a python version, boost is not compiled with. Another call afterwards with a supported python version returning success, but the target properties of the Boost::python<xy> target seems not to be complete then (the generated makefiles get some notfound libraries instead of the correct library. I don't know if this might have been fixed in versions of boost > 1.72. But it's definitely a bug in the cmake handling of boost, since the second call isn't going through the whole cmake code and returns immediately because of the existing Boost::python target (see: boost_python-config.cmake).

To cope with this, I've introduced the support to explicitly set the python versions to be checked from outside via PYTHON_BINDING_VERSIONS cmake variable. What you also did manually in the end above. See:
https://github.com/carla-simulator/map/pull/23/files#diff-4abfe42afb389def891c98e967838586R171

Hope now with the branch of the PR the following succeeds on your side:
cmake ../../ad_physics -DCMAKE_INSTALL_PREFIX=../../install/ad_physics -DBUILD_PYTHON_BINDING=ON -DPYTHON_BINDING_VERSIONS="3.7"

Could you please give it a try?
